lib/models/room.js

```diff
@@ -347,6 +354,45 @@ Room.prototype.recalculate = function(userId) {
     if (oldName !== this.name) {
         this.emit("Room.name", this);
     }
+
+
+
+    // recalculate read receipts, adding implicit ones where necessary
+    // NB. This is a duplication of logic for injecting implicit receipts,
+    // it would be technically possible to only ever generate these
+    // receipts in addEventsToTimeline but doing so means correctly
+    // choosing whether to keep or replace the existing receipt which
+    // is complex and slow. This is faster and more understandable.
+
+    var usersFound = {};
+    for (var i = this.timeline.length - 1; i >= 0; --i) {
+        // loop through the timeline backwards looking for either an
+        // event sent by each user or a real receipt from them.
+        // Replace the read receipt for that user with whichever
+        // occurs later in the timeline (ie. first because we're going
+        // backwards).
+        var e = this.timeline[i];
+
+        var readReceiptsForEvent = this.getReceiptsForEvent(e);
+
+        for (var receiptIt = 0; receiptIt < readReceiptsForEvent.length; ++receiptIt) {
```

Comment:
'receipt iterator' because I'd already used 'i' and I generally find using 'j' in such situations a recipe for accidentally referencing the wrong variable. Suggestions welcome.